### ISSUE TITLE
please provide borg-linuxold64 for borg-2.0.0 and beta releases

### ISSUE THREAD

**Issue by agrommek** [borgbackup/borg#7459]
## Have you checked borgbackup docs, FAQ, and open GitHub issues?

Yes

## Is this a BUG / ISSUE report or a QUESTION?

Question / polite request :-)

#### Your borg version (borg -V).

borg-2.0.0b5

#### Operating system (distribution) and version.

Synology NAS / DSM 7.1.1-42962 Update 4
Linux kernel 3.10.180 (--> old, but cannot be helped)
glibc 2.26, compiled with GCC 8.5.0 (also old, cannot be helped either)

#### Hardware / network configuration, and filesystems used.

ext4, BTRFS

## Describe the problem you're observing.

As the title says: Please continue to provide linux binaries for glibc < 2.28 (i.e. borg-linuxold64) for borg 2.0 and the beta releases.

#### Background

I run borg on a Synology NAS using the precompiled/packaged binaries. Worked great for the last 6 years or so. Right now I am in the process to modify my wrapper scripts for the new borg2 command line syntax. 

However, borg-linux64 does not run on my NAS, the glibc (version 2.26) appears to be too old. AFAIK there is no newer NAS-OS (Synology DSM) with a newer glibc. I cannot build borg-2.0.0 releases on the NAS because there are development headers (OpenSSL, probably others) missing and the pre-installed version of Python is too old anyway. There may be the option to cross-compile borg as a Synology package (instructions see [here](https://help.synology.com/developer-guide/)). But the developer guide appears to be targeted at people who know exactly what they do and are used to cross-compile stuff for different platforms - which does not include myself, unfortunately. Just fetching the binary from the releases page is far easier.

Please continue to provide borg-linuxold64 for borg-2.0 and beta releases.

Thanks for the great work!

**Comment by ThomasWaldmann** [comment]
the "linuxold" binaries for borg 1.2.x were made on a Debian stretch machine.

i tried adding the stretch machine to master branch, but it didn't work for building borg in master branch, so i removed it again. #7010

**Comment by ThomasWaldmann** [comment]
Hmm, guess one option could be to install xxhash header and lib from source. PRs or bounties welcome!

**Comment by bbx0** [comment]
Another option might be [staticx](https://staticx.readthedocs.io/en/latest/introduction.html). I was testing it as an alternative to the distroless images while looking into how to provide borg via docker. I decided against it, to ensure a fully scannable SBOM of the image content. Still, the build seemed to be fine.

```bash
staticx -l /lib/x86_64-linux-gnu/libgcc_s.so.1 --no-compress /pyinstaller/dist/borg.exe borg
```

I attached a build with staticx for borg 2.0.0b5 based on Debian 11 (without fuse support). [borg-2.0.0b5.tar.gz](https://github.com/borgbackup/borg/files/11045149/borg-2.0.0b5.tar.gz) This build is statically linked including glibc and should just work.

`libgcc_s.so.1` is required for borg2. A self-test (`borg debug info --debug`) works fine without it but `borg benchmark cpu` would fail, if not available.

(Statically linking all libs comes with security concerns as updating these requires to repackage the binary.)

**Comment by agrommek** [comment]
I really like the idea of 100% statically linked executables! 

Pro: 
- 100% self-contained
- no need to to provide several versions for compiled binaries (linuxold64, linux64, linuxnew64)
- no hassle with old distributions (Debian Stretch...) and new dependencies
- VM for building binaries can always be "up-to-date" (i.e. use newest version of Python - Python 3.11 is supposed to be faster in some usecases)

Con:
- Static linking might make binaries bigger (yet again)
- Statically linking libs comes with securtiy concerns (as mentioned by @bbx0)

The security concerns are IMHO valid. On the other hand, there are already lots of libs already bundled in the pyinstaller-produced binary. Adding a few more to make everything 100% standalone will probably not make a big difference?

I have no idea how the binaries are built today. If somebody were kind enough to point me the right direction, I would try to create a PR one of these days (I have seen https://github.com/borgbackup/borg/issues/7010, but I do not understand how building the binaries automatically works). 

That said, I think I re-invented the wheel. I managed to create a pyinstaller binary of borg-2.0.0b5 in a debian stretch docker container. 

Start a docker container like so:

`sudo docker run --mount type=bind,source=~/borg-output,target=/borg-output -ti debian:stretch-slim`

Then run the following command / the following script inside the container:

```
#!/bin/bash

# some path variables
BASEPATH="/root"
OUTPATH="/borg-output"

# install dependencies
cd ${BASEPATH}
apt-get update
apt-get -y upgrade

# install necessary tools
apt-get -y install wget unzip

# install build-tools
apt-get -y install build-essential udev

### openssl
OPENSSL='openssl-1.1.1t'

cd ${BASEPATH}
wget -O ${OPENSSL}.tar.gz https://www.openssl.org/source/${OPENSSL}.tar.gz
tar xvzf ${OPENSSL}.tar.gz
cd ${OPENSSL}
./config
make -j 14
make install
ldconfig
cd ${BASEPATH}

### attr
ATTR="attr-2.5.1"

cd ${BASEPATH}
wget -O ${ATTR}.tar.gz http://download.savannah.nongnu.org/releases/attr/${ATTR}.tar.gz
tar xvzf ${ATTR}.tar.gz
cd ${ATTR}
./configure
make -j 8
make install
ldconfig
cd ${BASEPATH}

### acl
ACL="acl-2.3.1"

cd ${BASEPATH}
wget -O ${ACL}.tar.gz http://download.savannah.nongnu.org/releases/acl/${ACL}.tar.gz
tar xvzf ${ACL}.tar.gz
cd ${ACL}
./configure
make -j 8
make install
ldconfig
cd ${BASEPATH}

### lz4
LZ4_NAME="lz4"
LZ4_VERSION="1.9.4"
LZ4=${LZ4_NAME}-${LZ4_VERSION}

cd ${BASEPATH}
wget -O ${LZ4}.tar.gz https://github.com/lz4/lz4/archive/refs/tags/v${LZ4_VERSION}.tar.gz
tar xvzf ${LZ4}.tar.gz
cd ${LZ4}
make -j 8
make install
ldconfig
cd ${BASEPATH}


### zstd
ZSTD_NAME="zstd"
ZSTD_VERSION="1.5.4"
ZSTD=${ZSTD_NAME}-${ZSTD_VERSION}

cd ${BASEPATH}
wget -O ${ZSTD}.tar.gz https://github.com/facebook/zstd/archive/refs/tags/v${ZSTD_VERSION}.tar.gz
tar xvzf ${ZSTD}.tar.gz
cd ${ZSTD}
make -j 8
make install
ldconfig
cd ${BASEPATH}


### xxHash
XXHASH_NAME="xxHash"
XXHASH_VERSION="0.8.1"
XXHASH=${XXHASH_NAME}-${XXHASH_VERSION}

cd ${BASEPATH}
wget -O ${XXHASH}.tar.gz https://github.com/Cyan4973/xxHash/archive/refs/tags/v${XXHASH_VERSION}.tar.gz
tar xvzf ${XXHASH}.tar.gz
cd ${XXHASH}
make -j 8
make install
ldconfig
cd ${BASEPATH}


### download ninja (binary)
# needed to build libfuse3

NINJA_VERSION="v1.11.1"
cd ${BASEPATH}
mkdir ninja
cd ninja
wget https://github.com/ninja-build/ninja/releases/download/${NINJA_VERSION}/ninja-linux.zip
unzip ninja-linux.zip
cp ninja /usr/local/bin
cd ${BASEPATH}


### installl python dependencies from debian repository
# leave out:
# libssl-dev --> compiled above
# libgdbm-compat-dev --> does not exist with stretch
cd ${BASEPATH}
apt-get -y install gdb lcov pkg-config libbz2-dev libffi-dev libgdbm-dev liblzma-dev libncurses5-dev libreadline6-dev libsqlite3-dev lzma lzma-dev tk-dev uuid-dev zlib1g-dev

# build Python
cd ${BASEPATH}
PYTHON="Python-3.10.10"
wget -O ${PYTHON}.tar.gz https://www.python.org/ftp/python/3.10.10/Python-3.10.10.tgz
tar xvzf ${PYTHON}.tar.gz
cd ${PYTHON}
./configure --enable-optimizations --enable-shared
make -j 14
make install
ldconfig
cd ${BASEPATH}

### prepare python, install dependencies for ninja and borg
/usr/local/bin/python3 -m pip install --upgrade pip pkgconfig wheel pyinstaller meson

### download and build libfuse3
FUSE="fuse-3.14.0"
FUSE_DIR="libfuse-${FUSE}"
cd ${BASEPATH}
wget -O ${FUSE}.tar.gz https://github.com/libfuse/libfuse/archive/refs/tags/${FUSE}.tar.gz
tar xvzf ${FUSE}.tar.gz
cd ${FUSE_DIR}
mkdir build
cd build
/usr/local/bin/meson setup ..
ninja
ninja install
ldconfig
cd ${BASEPATH}

# alternative when using llfuse
apt-get -y install libfuse-dev

# build borg backup
#/usr/local/bin/python3  -m pip install borgbackup==2.0.0b5
/usr/local/bin/python3 -m pip install --upgrade pyfuse3 llfuse
#/usr/local/bin/python3 -m pip install borgbackup[pyfuse3]==2.0.0b5
/usr/local/bin/python3 -m pip install borgbackup[llfuse]==2.0.0b5

# create single executable
BORG_BIN_NAME="borg-linuxold64-selfcompiled"
cd ${BASEPATH}
pyinstaller --clean --name ${BORG_BIN_NAME} --onefile /usr/local/bin/borg

# copy executable to output directory
cp ${BASEPATH}/dist/${BORG_BIN_NAME} ${OUTPATH}
```
This build many dependencies from source (openssl, attr, acl, lz4, zstd, xxHash, libfuse3, Python 3.10) and then borg itself.

The product of pyinstaller works on by NAS.

Maybe some of theses commands might help you to fix https://github.com/borgbackup/borg/issues/7010? And as I said, if somebody points me in the right direction, I could probably create a PR.

**Comment by ThomasWaldmann** [comment]
I don't know much about that low-level stuff, but I remember somebody saying the glibc should come from the installed OS and should match the kernel.

**Comment by bbx0** [comment]
I had a look at the Vagrantfile and and made a draft #7478 for it. There are some challenges with how the tox tests are invoked, when a borg.exe with a different FUSE_IMPL is in PATH but besides that, it seems fine. 

Generating a staticx binary in the process should be fairly easy. Maybe its worth testing if a `borg-staticx` variant (maybe for bullseye) causes issues. I presume there is at least a performance penalty for an unmatching glibc to kernel.

**Comment by ThomasWaldmann** [comment]
@bbx0 yes, the tox tests fail with the borg.exe binary if it does not match FUSE_IMPL, there is an open ticket about that...

Let's first finish this PR and do that staticx stuff after people had a chance to comment on the issue.

**Comment by ThomasWaldmann** [comment]
Fixed by #7478.
